terraform/transform_flatten.go

```diff
@@ -53,6 +53,12 @@ func (t *FlattenTransformer) Transform(g *Graph) error {
 
 		// Go through the subgraph and flatten all the nodes
 		for _, sv := range subgraph.Vertices() {
+			// If the vertex already has a subpath then we assume it has
+			// already been flattened. Ignore it.
+			if _, ok := sv.(GraphNodeSubPath); ok {
```

Comment:
I'm somewhat concerned that without something more directly expressive like `.(GraphNodeFlattened)` w/ `Flattened() bool` this would be easy to mess up elsewhere down the road.

But only somewhat. Not a dealbreaker, just food for thought.